ffi/src/std_impls.rs

```diff
@@ -1,64 +1,65 @@
 // Triggered by `&mut str` expansion
 #![allow(clippy::mut_mut, single_use_lifetimes)]
 
-use alloc::{
-    collections::{BTreeMap, BTreeSet},
-    string::String,
-    vec::Vec,
-};
+use alloc::{string::String, vec::Vec};
 use core::{mem::ManuallyDrop, ptr::NonNull};
 
 use crate::{
     ffi_type,
-    ir::InfallibleTransmute,
-    option::Niche,
     slice::{SliceMut, SliceRef},
     ReprC, WrapperTypeOf,
 };
 
 // NOTE: This can be contested as it is nowhere documented that String is
 // actually transmutable into Vec<u8>, but implicitly it should be
 // SAFETY: String type should be transmutable into Vec<u8>
-ffi_type! {unsafe impl Transparent for String[Vec<u8>] validated with {|target| core::str::from_utf8(target).is_ok()} }
+ffi_type! {
+    unsafe impl Transparent for String {
+        type Target = Vec<u8>;
+
+        validation_fn=unsafe {|target| core::str::from_utf8(target).is_ok()},
+        niche_value=SliceMut::null_mut()
+    }
+}
 // NOTE: `core::str::as_bytes` uses transmute internally which means that
 // even though it's a string slice it can be transmuted into byte slice.
-ffi_type! {unsafe impl<'slice> Transparent for &'slice str[&'slice [u8]] validated with {|target| core::str::from_utf8(target).is_ok()} }
+ffi_type! {
+    unsafe impl<'slice> Transparent for &'slice str {
+        type Target = &'slice [u8];
+
+        validation_fn=unsafe {|target| core::str::from_utf8(target).is_ok()},
+        niche_value=SliceRef::null()
+    }
+}
 #[cfg(feature = "non_robust_ref_mut")]
-ffi_type! {unsafe impl<'slice> Transparent for &'slice mut str[&'slice mut [u8]] validated with {|target| core::str::from_utf8(target).is_ok()} }
-ffi_type! {unsafe impl<T> Transparent for core::mem::ManuallyDrop<T>[T] validated with {|_| true} }
-ffi_type! {unsafe impl<T> Transparent for core::ptr::NonNull<T>[*mut T] validated with {|target: &*mut T| !target.is_null()} }
-ffi_type! {impl<K, V> Opaque for BTreeMap<K, V> }
-ffi_type! {impl<K> Opaque for BTreeSet<K> }
+ffi_type! {
+    unsafe impl<'slice> Transparent for &'slice mut str {
+        type Target = &'slice mut [u8];
+
+        validation_fn=unsafe {|target| core::str::from_utf8(target).is_ok()},
+        niche_value=SliceMut::null_mut()
+    }
+}
+ffi_type! {
+    unsafe impl<T> Transparent for core::mem::ManuallyDrop<T> {
+        type Target = T;
+    }
+}
+ffi_type! {
+    unsafe impl<T> Transparent for core::ptr::NonNull<T> {
+        type Target = *mut T;
+
+        validation_fn=unsafe {|target: &*mut T| !target.is_null()},
+        niche_value=core::ptr::null_mut()
+    }
+}
```

Comment:
this is the best API I could come up with for non-robust transparent items. Any ideas?

Reply:
The things I dislike:
* code in `validation_fn` is just plugged into `unsafe fn validate` method. This is bad because one can write any unsafe code and it will compile. I've added `validate_fn=unsafe{..}` marker to at least visually mark the unsafety 
* for robust transparent types the `ffi_type`doesn't require `validation_fn` because it's inferred to always be true. This is bad because user can forget to write a validation function if one is required

I've added appropriate description for the macro but I'm still concerned about potential misuses

Reply:
As far as I don't understand a half of code in `ffi` directory, I can't suggest anything meaningful

Reply:
My 2 cents: 

> ... is bad because one can write any unsafe code and it will compile.

Which is true of most `unsafe`. That's why people are discouraged from using `unsafe` unless they know what they're doing. It's just as bad as `eval`, but `eval` is only a problem when exposed at a prompt. 

We can't prevent all misuse, all we can do is provide safe alternatives, for typical cases. A few pseudo-marker traits (I believe you'd need to specify types) with verified sensible `validate` functions will do the trick. Doubt we can do much beyond that. 

> for robust transparent types the ffi_typedoesn't require validation_fn because it's inferred to always be true. This is bad because user can forget to write a validation function if one is required

As long as we have a tutorial that explains when a type is considered robust, this isn't our responsibility. In fact, I'll ask you this question: _who is the target audience for the FFI library_? People providing dynamic runtimes don't necessarily need to know the ins and outs of the system in order to build up an API. They don't need to know how to build a robust type; using primitive types _only_ is enough, and we have good coverage. We have more than sufficiently many types  to build a robust runtime, with some inefficiencies only apparent after years of usage and deployment in real projects. 

Supporting all possible types is a premature optimisation. But more to the point, we don't necessarily need to provide implementations directly. We have good examples with `ConstString` and a few other primitives, where we demonstrate the worst case scenarios and how to handle them. The people who are to be qualified to interact with the FFI are more than qualified to get what they need by looking at the code and at this point, writing a short tutorial and stabilising the FFI API would be all we need. 

The API you've come up with is reasonably easy to understand, a huge improvement over substrate, and sufficient for our purposes. What more do we need?